core/store/src/trie/shard_tries.rs

```diff
@@ -483,6 +483,36 @@ impl ShardTries {
             memtries.write().unwrap().delete_until_height(height);
         }
     }
+
+    /// Freezes in-memory trie for source shard and copies reference to it to
+    /// target shards.
+    /// Needed to serve queries for these shards just after resharding, before
+    /// proper memtries are loaded.
+    pub fn freeze_mem_tries(
+        &self,
+        source_shard_uid: ShardUId,
+        target_shard_uids: Vec<ShardUId>,
+    ) -> Result<(), StorageError> {
+        let mut outer_guard = self.0.mem_tries.write().unwrap();
+        let Some(memtries) = outer_guard.remove(&source_shard_uid) else {
+            return Err(StorageError::MemTrieLoadingError("Memtrie not loaded".to_string()));
+        };
+        let mut guard = memtries.write().unwrap();
+        let memtries = std::mem::replace(&mut *guard, MemTries::new(source_shard_uid));
+        let frozen_memtries = memtries.freeze();
+
+        for shard_uid in [vec![source_shard_uid], target_shard_uids].concat() {
```

Comment:
I'm curious, why do you need to also freeze the source shard uid? It makes sense by all means, I'm just trying to understand the logic.

Reply:
It's not "also" I think. Source shard uid memtrie always has to be frozen to be shared with children. Here we just put frozen memtrie back to the `ShardTries`, so that it can be used for proper fork handling.